Source/NSOrderedSet.m

```diff
@@ -0,0 +1,1763 @@
+      return;
+    }
+  i = [self count];
+  if (i > 0)
+    {
+      IMP	rem = 0;
+      IMP	get = [self methodForSelector: oaiSel];
+      BOOL	(*eq)(id, SEL, id)
+	= (BOOL (*)(id, SEL, id))[anObject methodForSelector: eqSel];
+
+      while (i-- > 0)
+	{
+	  id	o = (*get)(self, oaiSel, i);
+
+	  if (o == anObject || (*eq)(anObject, eqSel, o) == YES)
+	    {
+	      if (rem == 0)
+		{
+		  rem = [self methodForSelector: remSel];
+		  /*
+		   * We need to retain the object so that when we remove the
+		   * first equal object we don't get left with a bad object
+		   * pointer for later comparisons.
+		   */
+		  RETAIN(anObject);
+		}
+	      (*rem)(self, remSel, i);
+	    }
+	}
+      if (rem != 0)
+	{
+	  RELEASE(anObject);
+	}
+    }
+}
+
+- (void)removeObjectAtIndex:(NSUInteger)index  // required override
+{
+  [self subclassResponsibility: _cmd];
+}
+
+- (void) _removeObjectsFromIndices: (NSUInteger*)indices
+			numIndices: (NSUInteger)count
+{
+  if (count > 0)
+    {
+      NSUInteger	to = 0;
+      NSUInteger	from = 0;
+      NSUInteger	i;
+      GS_BEGINITEMBUF(sorted, count, NSUInteger);
+
+      while (from < count)
+	{
+	  NSUInteger	val = indices[from++];
+
+	  i = to;
+	  while (i > 0 && sorted[i-1] > val)
+	    {
+	      i--;
+	    }
+	  if (i == to)
+	    {
+	      sorted[to++] = val;
+	    }
+	  else if (sorted[i] != val)
+	    {
+	      NSUInteger	j = to++;
+
+	      if (sorted[i] < val)
+		{
+		  i++;
+		}
+	      while (j > i)
+		{
+		  sorted[j] = sorted[j-1];
+		  j--;
+		}
+	      sorted[i] = val;
+	    }
+	}
+
+      if (to > 0)
+	{
+	  IMP	rem = [self methodForSelector: remSel];
+
+	  while (to--)
+	    {
+	      (*rem)(self, remSel, sorted[to]);
+	    }
+	}
+      GS_ENDITEMBUF();
+    }
+}
+
+- (void)removeObjectsAtIndexes:(NSIndexSet *)indexes
+{
+  NSUInteger count = [indexes count];
+  NSUInteger indexArray[count];
+
+  [indexes getIndexes: indexArray
+             maxCount: count
+         inIndexRange: NULL];
+
+  [self _removeObjectsFromIndices: indexArray
+		       numIndices: count];
+}
+
+- (void)removeObjectsInArray:(NSArray *)otherArray
+{
+  NSUInteger	c = [otherArray count];
+
+  if (c > 0)
+    {
+      NSUInteger	i;
+      IMP	get = [otherArray methodForSelector: oaiSel];
+      IMP	rem = [self methodForSelector: @selector(removeObject:)];
+
+      for (i = 0; i < c; i++)
+	(*rem)(self, @selector(removeObject:), (*get)(otherArray, oaiSel, i));
+    }
+}
+
+- (void)removeObjectsInRange:(NSRange)aRange
+{
+  NSUInteger	i;
+  NSUInteger	s = aRange.location;
+  NSUInteger	c = [self count];
+
+  i = aRange.location + aRange.length;
+
+  if (c < i)
+    i = c;
+
+  if (i > s)
+    {
+      IMP	rem = [self methodForSelector: remSel];
+
+      while (i-- > s)
+	{
+	  (*rem)(self, remSel, i);
+	}
+    }
+}
+
+- (void)removeAllObjects
+{
+  NSUInteger	c = [self count];
+
+  if (c > 0)
+    {
+      IMP	remLast = [self methodForSelector: rlSel];
+
+      while (c--)
+	{
```

Comment:
This looks like you are sorting the indices array, but it is already sorted, which make the whole method superfluouses.

Reply:
This is how this seems to be done in NSArray for whatever reason.

Reply:
Ah, here it is! The NSArray method is a (deprecated) public method. There the indexes could come in an arbitrary order. Here this is not the case.